File: patterns/backtracking/7. Permutations II/7. Permutations II.js
/**
 * Given a collection of numbers, nums, that might contain duplicates,
 * return all the possible unique permutations in any order.
 */

function permuteUnique(nums) {
    const result = [];

    function backtrack(start) {
        if (start === nums.length) {
            result.push(nums.slice());
            return;
        }

        const seen = new Set();

        for (let i = start; i < nums.length; i++) {
            //to avoid duplicate pairs
            if (!seen.has(nums[i])) {
                seen.add(nums[i]);

                [nums[i], nums[start]] = [nums[start], nums[i]];
                backtrack(start + 1);
                [nums[i], nums[start]] = [nums[start], nums[i]];
            }
        }
    }

    backtrack(0);
    return result;
}

// Example usage
const nums = [1, 1, 2];
console.log(permuteUnique(nums)); // [[1,1,2],[1,2,1],[2,1,1]]

/*
Time complexity: O(n! * n)
Space complexity: O(n)
where n is the length of the input array.
*/

File: patterns/backtracking/7. Permutations II/7. Permutations II.py
# 47. Permutations II
# LeetCode Problem: https://leetcode.com/problems/permutations-ii/

def permuteUnique(nums):
    """
    Given a collection of numbers, nums, that might contain duplicates, 
    return all the possible unique permutations in any order.
    """
    result = []
    nums.sort()  # Sort to handle duplicates
    
    def backtrack(current_permutation, used):
        if len(current_permutation) == len(nums):
            result.append(current_permutation[:])
            return
        
        for i in range(len(nums)):
            if used[i] or (i > 0 and nums[i] == nums[i-1] and not used[i-1]):
                continue
            
            used[i] = True
            current_permutation.append(nums[i])
            backtrack(current_permutation, used)
            current_permutation.pop()
            used[i] = False
    
    backtrack([], [False] * len(nums))
    return result

# Example usage
if __name__ == "__main__":
    nums = [1, 1, 2]
    print(permuteUnique(nums))  # [[1,1,2],[1,2,1],[2,1,1]]